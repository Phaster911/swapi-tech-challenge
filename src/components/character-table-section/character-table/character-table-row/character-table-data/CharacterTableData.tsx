import styled from "styled-components";
import { controlBorderRadius, QUERIES } from "../../../../../style-constants";
import { ReactNode } from "react";

const StyledTableData = styled.td`
  --radius: calc(${controlBorderRadius} * 2);
  padding: 0.5rem;
  display: grid;
  grid-template-columns: 15ch auto;
  gap: 0.5rem;
  text-align: left;

  &:first-child {
    border-top-left-radius: var(--radius);
    border-top-right-radius: var(--radius);
  }
  &:last-child {
    border-bottom-left-radius: var(--radius);
    border-bottom-right-radius: var(--radius);
  }

  &::before {
    content: attr(data-cell) ": ";
    text-transform: capitalize;
    font-weight: 700;
  }

  @media ${QUERIES.tabletAndUp} {
    display: revert;
    padding: 1rem 0.5rem;

    &::before {
      display: none;
    }

    &:first-child {
      border-top-left-radius: var(--radius);
      border-bottom-left-radius: var(--radius);
      border-top-right-radius: 0;
    }
    &:last-child {
      border-top-right-radius: var(--radius);
      border-bottom-right-radius: var(--radius);
      border-bottom-left-radius: 0;
    }
  }
`;

export type CharacterTableDataProps = {
  cellName: string;
  children: ReactNode;
};

export default function CharacterTableData({
  cellName,
  children,
}: CharacterTableDataProps) {
  return <StyledTableData data-cell={cellName}>{children}</StyledTableData>;
}
