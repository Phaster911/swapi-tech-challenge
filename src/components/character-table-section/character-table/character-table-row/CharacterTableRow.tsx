import styled from "styled-components";
import CharacterTableData from "./character-table-data/CharacterTableData";
import { Person } from "../../../../react-query/queries/use-get-people.hook";

const StyledTableRow = styled.tr`
  &:nth-child(even) td {
    background-color: #f0f0f0;
    overflow: hidden;
  }

  &:last-child {
    border-bottom: revert;
  }
`;

export type CharacterTableRowProps = Omit<Person, "id">;

export default function CharacterTableRow({
  name,
  gender,
  birthYear,
  mass,
  filmsEntered,
  height,
}: CharacterTableRowProps) {
  return (
    <StyledTableRow data-testid="table-row">
      <CharacterTableData cellName="name">{name}</CharacterTableData>
      <CharacterTableData cellName="gender">{gender}</CharacterTableData>
      <CharacterTableData cellName="birth year">{birthYear}</CharacterTableData>
      <CharacterTableData cellName="height">{height}</CharacterTableData>
      <CharacterTableData cellName="mass">{mass}</CharacterTableData>
      <CharacterTableData cellName="films entered">
        {filmsEntered}
      </CharacterTableData>
    </StyledTableRow>
  );
}
