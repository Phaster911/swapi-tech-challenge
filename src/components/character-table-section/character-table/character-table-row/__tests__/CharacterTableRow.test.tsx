import { render } from "@testing-library/react";
import CharacterTableRow, {
  CharacterTableRowProps,
} from "../CharacterTableRow";
import { Person } from "../../../../../react-query/queries/use-get-people.hook";
import { CharacterTableDataProps } from "../character-table-data/CharacterTableData";

const mockedCharacterTableData = vi
  .fn()
  .mockReturnValue(<td>CharacterTableData</td>);
vi.mock("../character-table-data/CharacterTableData", () => ({
  default: (props: CharacterTableDataProps) => mockedCharacterTableData(props),
}));

const person: Person = {
  id: "1",
  name: "name-1",
  gender: "gender-1",
  birthYear: "year-1",
  mass: "mass-1",
  filmsEntered: 0,
  height: "height-1",
};

describe("<CharacterTableRow />", () => {
  let props: CharacterTableRowProps;
  beforeEach(() => {
    props = {
      ...person,
    };
  });
  it("renders the component", () => {
    const { container } = render(
      <table>
        <tbody>
          <CharacterTableRow {...props} />
        </tbody>
      </table>
    );
    expect(container).toMatchSnapshot();
    expect(mockedCharacterTableData).toHaveBeenCalledTimes(6);
    expect(mockedCharacterTableData).toHaveBeenNthCalledWith(1, {
      cellName: "name",
      children: "name-1",
    });
    expect(mockedCharacterTableData).toHaveBeenNthCalledWith(2, {
      cellName: "gender",
      children: "gender-1",
    });
    expect(mockedCharacterTableData).toHaveBeenNthCalledWith(3, {
      cellName: "birth year",
      children: "year-1",
    });
    expect(mockedCharacterTableData).toHaveBeenNthCalledWith(4, {
      cellName: "height",
      children: "height-1",
    });
    expect(mockedCharacterTableData).toHaveBeenNthCalledWith(5, {
      cellName: "mass",
      children: "mass-1",
    });
    expect(mockedCharacterTableData).toHaveBeenNthCalledWith(6, {
      cellName: "films entered",
      children: 0,
    });
  });
});
