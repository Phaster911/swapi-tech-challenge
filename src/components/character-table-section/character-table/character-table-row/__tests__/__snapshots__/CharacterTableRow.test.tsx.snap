// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CharacterTableRow /> > renders the component 1`] = `
.c0:nth-child(even) td {
  background-color: #f0f0f0;
  overflow: hidden;
}

.c0:last-child {
  border-bottom: revert;
}

<div>
  <table>
    <tbody>
      <tr
        class="c0"
        data-testid="table-row"
      >
        <td>
          CharacterTableData
        </td>
        <td>
          CharacterTableData
        </td>
        <td>
          CharacterTableData
        </td>
        <td>
          CharacterTableData
        </td>
        <td>
          CharacterTableData
        </td>
        <td>
          CharacterTableData
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
